File: Stack/GFG/Medium/InfixtoPostfix/InfixtoPostfix.java
// https://www.geeksforgeeks.org/problems/infix-to-postfix-1587115620/1

import java.util.ArrayDeque;
import java.util.Deque;

public class InfixtoPostfix {
    public static String infixToPostfix(String s) {
        // Your code here
        Deque<Character> st = new ArrayDeque<>();
        StringBuilder post = new StringBuilder("");
        for(int i=0; i<s.length(); i++) {
            char ch = s.charAt(i);
            if(Character.isLetterOrDigit(ch)) {
                post.append(ch);
            } else if(ch == '(') {
                st.push(ch);
            } else if(ch == ')') {
                while(st.peek() != '(') {
                    post.append(st.pop());
                }
                st.pop();
            } else {
                while(!st.isEmpty() && priority(ch) <= priority(st.peek())) {
                    post.append(st.pop());
                }
                st.push(ch);
            }
        }
        while(!st.isEmpty()) {
            post.append(st.pop());
        }
        return post.toString();
    }
    static int priority(char ch) {
        switch(ch) {
            case '^': return 2;
            case '*': case '/': return 1;
            case '+': case '-': return 0;
        }
        return -1;
    }
}

File: Stack/GFG/Medium/InfixtoPostfix/InfixtoPostfix.cpp
// https://www.geeksforgeeks.org/problems/infix-to-postfix-1587115620/1
#include<bits/stdc++.h>
using namespace std;
string infixToPostfix(string &s)
{
    // Your code here
    stack<char> st;
    string post = "";
    for (char &ch : s)
    {
        if (isalnum(ch))
        {
            post += ch;
        }
        else if (ch == '(')
        {
            st.push(ch);
        }
        else if (ch == ')')
        {
            while (st.top() != '(')
            {
                post += st.top();
                st.pop();
            }
            st.pop();
        }
        else
        {
            while (!st.empty() && priority(ch) <= priority(st.top()))
            {
                post += st.top();
                st.pop();
            }
            st.push(ch);
        }
    }
    while (!st.empty())
    {
        post += st.top();
        st.pop();
    }
    return post;
}
int priority(char &ch)
{
    switch (ch)
    {
    case '^':
        return 2;
    case '*':
    case '/':
        return 1;
    case '+':
    case '-':
        return 0;
    }
    return -1;
}